### PR TITLE
1.20.3-1.20.4 protocol

### PR DESCRIPTION
**Pull request by wgaylord**
This protocol file consists of adding 8 new packets. and removing 2, updating a few packets as well.

Each packet edit was commit as its own commit to allow seeing what changed. 

Also at the end I went thru and updated each place where the use of a string had been replaced with NBT  This was done by looking at the list on https://wiki.vg/index.php?title=Pre-release_protocol&oldid=18635 and checking the MC source to see which string field in that packet had changed since some have stayed as normal string while other now use NBT.

### CODE REVIEW

**Comment by wgaylord** [comment]
This also requires https://github.com/PrismarineJS/prismarine-nbt/pull/79/files since from 1.20.2 all NBT over network is like this.

**Comment by rom1504** [comment]
No change in 1.20.4 ?

**Comment by wgaylord** [comment]
Nope, 1.20.3 and 1.20.4 are protocol compatable. (.4 was a 1 line fix to a
world corrruption bug)

**Comment by wgaylord** [comment]
Okay, I think I have everything done, went thru every single packet, there is possibly an issue with the update_objective packet. Since I don't think how I am decoding it is the best way but I couldn't think of any other way. (I don't even know if what I tied will actually work.)

So requesting someone look over that packet.

**Comment by rom1504** [comment]
diff between 1.20.2 and 1.20.3 protocol.json file:

**Comment by rom1504** [comment]
```
 diff data/pc/1.20.2/protocol.json data/pc/1.20.3/protocol.json 
225c225
<           "25": [
---
>           "27": [
234c234
<           "31": [
---
>           "33": [
243c243
<           "40": [
---
>           "42": [
252c252
<           "41": [
---
>           "43": [
304c304
<           "93": [
---
>           "96": [
418c418
<           "component": "string",
---
>           "component": "anonymousNbt",
421c421
<             "string"
---
>             "anonymousNbt"
1710c1710
<               "type": "string"
---
>               "type": "anonymousNbt"
1745c1745
<         "packet_resource_pack_send": [
---
>         "packet_remove_resource_pack": [
1747c1747,1761
<           [
---
>           [{
>               "name": "uuid",
>               "type": [
>                 "option",
>                 "UUID"
>               ]
>             }
>           ]
>         ],
>         "packet_add_resource_pack": [
>           "container",
>           [{
>               "name": "uuid",
>               "type": "UUID"
>             },
1764c1778
<                 "string"
---
>                 "anonymousNbt"
1827,1829c1841,1844
<                     "0x06": "resource_pack_send",
<                     "0x07": "feature_flags",
<                     "0x08": "tags"
---
>                     "0x06": "remove_resource_pack",
>                     "0x07": "add_resource_pack",
>                     "0x08": "feature_flags",
>                     "0x09": "tags"
1847c1862,1863
<                     "resource_pack_send": "packet_resource_pack_send",
---
>                     "remove_resource_pack": "packet_remove_resource_pack",
>                     "add_resource_pack": "packet_add_resource_pack",
1934c1950,1953
<           [
---
>           [{
>               "name": "uuid",
>               "type": "UUID"
>             },
2153c2172
<                                       "type": "string"
---
>                                       "type": "anonymousNbt"
2157c2176
<                                       "type": "string"
---
>                                       "type": "anonymousNbt"
2389,2390c2408,2409
<                     "0": "string",
<                     "3": "string"
---
>                     "0": "anonymousNbt",
>                     "3": "anonymousNbt"
2512c2531
<                           "string"
---
>                           "anonymousNbt"
2584c2603
<                     "true": "string"
---
>                     "true": "varint"
2665c2684
<               "type": "string"
---
>               "type": "anonymousNbt"
2811c2830
<               "type": "string"
---
>               "type": "anonymousNbt"
2820c2839
<               "type": "string"
---
>               "type": "anonymousNbt"
2828c2847
<               "type": "string"
---
>               "type": "anonymousNbt"
2834,2835c2853,2854
<                 "string"
<               ]
---
>                 "anonymousNbt"
>               ]  
2907a2927,2949
>             },
>             {
>               "name": "block_interaction_type",
>               "type": "varint"
>             },
>             {
>               "name": "small_explosion_particle",
>               "type": "particle"
>             },
>             {
>               "name": "large_explosion_particle",
>               "type": "particle"
>             },
>             {
>               "name": "explosion_sound_name",
>               "type": "string"
>             },
>             {
>               "name": "sound_range",
>               "type": [
>                 "option",
>                 "f32"
>               ]
3375c3417
<                             "string"
---
>                             "anonymousNbt"
3725c3767
<                 "string"
---
>                 "anonymousNbt"
3757c3799
<               "type": "string"
---
>               "type": "anonymousNbt"
3763c3805
<                 "string"
---
>                 "anonymousNbt"
3790c3832
<               "type": "string"
---
>               "type": "anonymousNbt"
4058c4100
<                                 "string"
---
>                                 "anonymousNbt"
4062c4104
<                                 "string"
---
>                                 "anonymousNbt"
4066c4108
<                                 "string"
---
>                                 "anonymousNbt"
4070c4112
<                                 "string"
---
>                                 "anonymousNbt"
4074c4116
<                                 "string"
---
>                                 "anonymousNbt"
4078c4120
<                                 "string"
---
>                                 "anonymousNbt"
4082c4124
<                                 "string"
---
>                                 "anonymousNbt"
4086c4128
<                                 "string"
---
>                                 "anonymousNbt"
4090c4132
<                                 "string"
---
>                                 "anonymousNbt"
4094c4136
<                                 "string"
---
>                                 "anonymousNbt"
4098c4140
<                                 "string"
---
>                                 "anonymousNbt"
4102c4144
<                                 "string"
---
>                                 "anonymousNbt"
4106c4148
<                                 "string"
---
>                                 "anonymousNbt"
4110c4152
<                                 "string"
---
>                                 "anonymousNbt"
4114c4156
<                                 "string"
---
>                                 "anonymousNbt"
4118c4160
<                                 "string"
---
>                                 "anonymousNbt"
4122c4164
<                                 "string"
---
>                                 "anonymousNbt"
4126c4168
<                                 "string"
---
>                                 "anonymousNbt"
4130c4172
<                                 "string"
---
>                                 "anonymousNbt"
4134c4176
<                                 "string"
---
>                                 "anonymousNbt"
4138c4180
<                                 "string"
---
>                                 "anonymousNbt"
4142c4184
<                                 "string"
---
>                                 "anonymousNbt"
4146c4188
<                                 "string"
---
>                                 "anonymousNbt"
4150c4192
<                                 "string"
---
>                                 "anonymousNbt"
4154c4196
<                                 "string"
---
>                                 "anonymousNbt"
4158c4200
<                                 "string"
---
>                                 "anonymousNbt"
4162c4204
<                                 "string"
---
>                                 "anonymousNbt"
4166c4208
<                                 "string"
---
>                                 "anonymousNbt"
4170c4212
<                                 "string"
---
>                                 "anonymousNbt"
4174c4216
<                                 "string"
---
>                                 "anonymousNbt"
4178c4220
<                                 "string"
---
>                                 "anonymousNbt"
4182c4224
<                                 "string"
---
>                                 "anonymousNbt"
4327c4369,4398
<         "packet_resource_pack_send": [
---
>         "packet_reset_score": [
>           "container",
>           [
>             {
>               "name": "entity_name",
>               "type": "string"
>             },
>             {
>               "name": "objective_name",
>               "type": [
>                 "option",
>                 "string"
>               ]
>             }
>           
>           ]
>         ],
>         "packet_remove_resource_pack": [
>           "container",
>           [
>             {
>               "name": "uuid",
>               "type": [
>                 "option",
>                 "UUID"
>               ]
>             }
>           ]
>         ],
>         "packet_add_resource_pack": [
4330a4402,4405
>               "name": "uuid",
>               "type": "UUID"
>             },
>             {
4346c4421
<                 "string"
---
>                 "anonymousNbt"
4606,4607c4681,4682
<                     "0": "string",
<                     "2": "string"
---
>                     "0": "anonymousNbt",
>                     "2": "anonymousNbt"
4625a4701,4754
>             },
>             {
>               "name": "number_format",
>               "type": [
>                 "switch",
>                 {
>                   "compareTo": "action",
>                   "fields": {
>                     "0": [
>                       "option",
>                       "varint"
>                     ],
>                     "2":[
>                       "option",
>                       "varint"
>                     ]
>                   },
>                   "default": "void"
>                 }
>               ]
>             },
>             {
>               "name": "styling",
>               "type": [
>                 "switch",
>                 {
>                   "compareTo": "action",
>                   "fields": {
>                     "0": [
>                       "switch",
>                       {
>                         "compareTo": "number_format",
>                         "fields": {
>                           "1": "anonymousNbt",
>                           "2": "anonymousNbt"
>                         },
>                         "default": "void"
>                       }
>                     ],
>                     "2":[
>                       "switch",
>                       {
>                         "compareTo": "number_format",
>                         "fields": {
>                           "1": "anonymousNbt",
>                           "2": "anonymousNbt"
>                         },
>                         "default": "void"
>                       }
>                     ]
>                   },
>                   "default": "void"
>                 }
>               ]
4666,4667c4795,4796
<                     "0": "string",
<                     "2": "string"
---
>                     "0": "anonymousNbt",
>                     "2": "anonymousNbt"
4736,4737c4865,4866
<                     "0": "string",
<                     "2": "string"
---
>                     "0": "anonymousNbt",
>                     "2": "anonymousNbt"
4750,4751c4879,4880
<                     "0": "string",
<                     "2": "string"
---
>                     "0": "anonymousNbt",
>                     "2": "anonymousNbt"
4800,4803d4928
<               "name": "action",
<               "type": "varint"
<             },
<             {
4808a4934,4937
>               "type": "varint"
>             },
>             {
>               "name": "display_name",
4810,4817c4939,4947
<                 "switch",
<                 {
<                   "compareTo": "action",
<                   "fields": {
<                     "1": "void"
<                   },
<                   "default": "varint"
<                 }
---
>                 "option",
>                 "anonymousNbt"
>               ]
>             },
>             {
>               "name": "number_format",
>               "type": [
>                 "option",
>                 "varint"
5012c5142
<               "type": "string"
---
>               "type": "anonymousNbt"
5025c5155
<               "type": "string"
---
>               "type": "anonymousNbt"
5029c5159
<               "type": "string"
---
>               "type": "anonymousNbt"
5189c5319
<               "type": "string"
---
>               "type": "anonymousNbt"
5265c5395
<                                 [
---
>                                 [                                  
5267,5268c5397,5398
<                                     "name": "width",
<                                     "type": "varint"
---
>                                     "name": "group",
>                                     "type": "string"
5271c5401
<                                     "name": "height",
---
>                                     "name": "category",
5275,5276c5405,5406
<                                     "name": "group",
<                                     "type": "string"
---
>                                     "name": "width",
>                                     "type": "varint"
5279c5409
<                                     "name": "category",
---
>                                     "name": "height",
5480c5610
<               "type": "string"
---
>               "type": "anonymousNbt"
5564c5694
<               "type": "string"
---
>               "type": "anonymousNbt"
5573c5703
<               "type": "string"
---
>               "type": "anonymousNbt"
5679a5810,5831
>         "packet_set_ticking_state": [
>           "container",
>           [
>            {
>               "name": "tick_rate",
>               "type": "f32"
>             },
>             {
>               "name": "is_frozen",
>               "type": "bool"
>             }
>           ]
>         ],
>         "packet_step_tick": [
>           "container",
>           [
>            {
>               "name": "tick_steps",
>               "type": "varint"
>             }
>           ]
>         ],
5703c5855
<                     "0x0d": "chunk_batch_start",
---
>                     "0x0d": "chunk_batch_start",                    
5756,5802c5908,5958
<                     "0x42": "resource_pack_send",
<                     "0x43": "respawn",
<                     "0x44": "entity_head_rotation",
<                     "0x45": "multi_block_change",
<                     "0x46": "select_advancement_tab",
<                     "0x47": "server_data",
<                     "0x48": "action_bar",
<                     "0x49": "world_border_center",
<                     "0x4a": "world_border_lerp_size",
<                     "0x4b": "world_border_size",
<                     "0x4c": "world_border_warning_delay",
<                     "0x4d": "world_border_warning_reach",
<                     "0x4e": "camera",
<                     "0x4f": "held_item_slot",
<                     "0x50": "update_view_position",
<                     "0x51": "update_view_distance",
<                     "0x52": "spawn_position",
<                     "0x53": "scoreboard_display_objective",
<                     "0x54": "entity_metadata",
<                     "0x55": "attach_entity",
<                     "0x56": "entity_velocity",
<                     "0x57": "entity_equipment",
<                     "0x58": "experience",
<                     "0x59": "update_health",
<                     "0x5a": "scoreboard_objective",
<                     "0x5b": "set_passengers",
<                     "0x5c": "teams",
<                     "0x5d": "scoreboard_score",
<                     "0x5e": "simulation_distance",
<                     "0x5f": "set_title_subtitle",
<                     "0x60": "update_time",
<                     "0x61": "set_title_text",
<                     "0x62": "set_title_time",
<                     "0x63": "entity_sound_effect",
<                     "0x64": "sound_effect",
<                     "0x65": "start_configuration",
<                     "0x66": "stop_sound",
<                     "0x67": "system_chat",
<                     "0x68": "playerlist_header",
<                     "0x69": "nbt_query_response",
<                     "0x6a": "collect",
<                     "0x6b": "entity_teleport",
<                     "0x6c": "advancements",
<                     "0x6d": "entity_update_attributes",
<                     "0x6e": "entity_effect",
<                     "0x6f": "declare_recipes",
<                     "0x70": "tags"
---
>                     "0x42": "reset_score",
>                     "0x43": "remove_resource_pack",
>                     "0x44": "add_resource_pack",
>                     "0x45": "respawn",
>                     "0x46": "entity_head_rotation",
>                     "0x47": "multi_block_change",
>                     "0x48": "select_advancement_tab",
>                     "0x49": "server_data",
>                     "0x4a": "action_bar",
>                     "0x4b": "world_border_center",
>                     "0x4c": "world_border_lerp_size",
>                     "0x4d": "world_border_size",
>                     "0x4e": "world_border_warning_delay",
>                     "0x4f": "world_border_warning_reach",
>                     "0x50": "camera",
>                     "0x51": "held_item_slot",
>                     "0x52": "update_view_position",
>                     "0x53": "update_view_distance",
>                     "0x54": "spawn_position",
>                     "0x55": "scoreboard_display_objective",
>                     "0x56": "entity_metadata",
>                     "0x57": "attach_entity",
>                     "0x58": "entity_velocity",
>                     "0x59": "entity_equipment",
>                     "0x5a": "experience",
>                     "0x5b": "update_health",
>                     "0x5c": "scoreboard_objective",
>                     "0x5d": "set_passengers",
>                     "0x5e": "teams",
>                     "0x5f": "scoreboard_score",
>                     "0x60": "simulation_distance",
>                     "0x61": "set_title_subtitle",
>                     "0x62": "update_time",
>                     "0x63": "set_title_text",
>                     "0x64": "set_title_time",
>                     "0x65": "entity_sound_effect",
>                     "0x66": "sound_effect",
>                     "0x67": "start_configuration",
>                     "0x68": "stop_sound",
>                     "0x69": "system_chat",
>                     "0x6a": "playerlist_header",
>                     "0x6b": "nbt_query_response",
>                     "0x6c": "collect",
>                     "0x6d": "entity_teleport",
>                     "0x6e": "set_ticking_state",
>                     "0x6f": "step_tick",
>                     "0x70": "advancements",
>                     "0x71": "entity_update_attributes",
>                     "0x72": "entity_effect",
>                     "0x73": "declare_recipes",
>                     "0x74": "tags"
5880c6036,6038
<                     "resource_pack_send": "packet_resource_pack_send",
---
>                     "reset_score": "packet_reset_score",
>                     "remove_resource_pack": "packet_remove_resource_pack",
>                     "add_resource_pack": "packet_add_resource_pack",
5926c6084,6086
<                     "tags": "packet_tags"
---
>                     "tags": "packet_tags",
>                     "set_ticking_state": "packet_set_ticking_state",
>                     "step_tick": "packet_step_tick"
6742a6903,6906
>               "name": "uuid",
>               "type": "UUID"
>             },
>             {
6795a6960,6967
>             },
>             {
>               "name": "selection_priority",
>               "type": "varint"
>             },
>             {
>               "name": "placement_priority",
>               "type": "varint"
6975,6976c7147,7165
<               "name": "id",
<               "type": "i64"
---
>                 "name": "id",
>                 "type": "i64"
>             }
>           ]
>         ],
>         "packet_set_slot_state": [
>           "container",
>           [
>             {
>                 "name": "slot_id",
>                 "type": "varint"
>             },
>             {
>                 "name": "window_id",
>                 "type": "varint"
>             },
>             {
>                 "name": "state",
>                 "type": "bool"
7005,7043c7194,7233
<                     "0x0f": "custom_payload",
<                     "0x10": "edit_book",
<                     "0x11": "query_entity_nbt",
<                     "0x12": "use_entity",
<                     "0x13": "generate_structure",
<                     "0x14": "keep_alive",
<                     "0x15": "lock_difficulty",
<                     "0x16": "position",
<                     "0x17": "position_look",
<                     "0x18": "look",
<                     "0x19": "flying",
<                     "0x1a": "vehicle_move",
<                     "0x1b": "steer_boat",
<                     "0x1c": "pick_item",
<                     "0x1d": "ping_request",
<                     "0x1e": "craft_recipe_request",
<                     "0x1f": "abilities",
<                     "0x20": "block_dig",
<                     "0x21": "entity_action",
<                     "0x22": "steer_vehicle",
<                     "0x23": "pong",
<                     "0x24": "recipe_book",
<                     "0x25": "displayed_recipe",
<                     "0x26": "name_item",
<                     "0x27": "resource_pack_receive",
<                     "0x28": "advancement_tab",
<                     "0x29": "select_trade",
<                     "0x2a": "set_beacon_effect",
<                     "0x2b": "held_item_slot",
<                     "0x2c": "update_command_block",
<                     "0x2d": "update_command_block_minecart",
<                     "0x2e": "set_creative_slot",
<                     "0x2f": "update_jigsaw_block",
<                     "0x30": "update_structure_block",
<                     "0x31": "update_sign",
<                     "0x32": "arm_animation",
<                     "0x33": "spectate",
<                     "0x34": "block_place",
<                     "0x35": "use_item"
---
>                     "0x0f": "set_slot_state",
>                     "0x10": "custom_payload",
>                     "0x11": "edit_book",
>                     "0x12": "query_entity_nbt",
>                     "0x13": "use_entity",
>                     "0x14": "generate_structure",
>                     "0x15": "keep_alive",
>                     "0x16": "lock_difficulty",
>                     "0x17": "position",
>                     "0x18": "position_look",
>                     "0x19": "look",
>                     "0x1a": "flying",
>                     "0x1b": "vehicle_move",
>                     "0x1c": "steer_boat",
>                     "0x1d": "pick_item",
>                     "0x1e": "ping_request",
>                     "0x1f": "craft_recipe_request",
>                     "0x20": "abilities",
>                     "0x21": "block_dig",
>                     "0x22": "entity_action",
>                     "0x23": "steer_vehicle",
>                     "0x24": "pong",
>                     "0x25": "recipe_book",
>                     "0x26": "displayed_recipe",
>                     "0x27": "name_item",
>                     "0x28": "resource_pack_receive",
>                     "0x29": "advancement_tab",
>                     "0x2a": "select_trade",
>                     "0x2b": "set_beacon_effect",
>                     "0x2c": "held_item_slot",
>                     "0x2d": "update_command_block",
>                     "0x2e": "update_command_block_minecart",
>                     "0x2f": "set_creative_slot",
>                     "0x30": "update_jigsaw_block",
>                     "0x31": "update_structure_block",
>                     "0x32": "update_sign",
>                     "0x33": "arm_animation",
>                     "0x34": "spectate",
>                     "0x35": "block_place",
>                     "0x36": "use_item"
7108c7298,7299
<                     "ping_request": "packet_ping_request"
---
>                     "ping_request": "packet_ping_request",
>                     "set_slot_state": "packet_set_slot_state"
7118c7309
< }
\ No newline at end of file
---
> }
```

**Comment by rom1504** [comment]
better way to visualize the diff https://github.com/PrismarineJS/minecraft-data/compare/diff_1.20.3?expand=1

**Comment by rom1504** [comment]
I'll review but if you want to take a look, it'll help @extremeheat

File: data/pc/1.20.3/protocol.json [hunk 1]
```diff
@@ -0,0 +1,7309 @@
...
+              "name": "small_explosion_particle",
+              "type": "particle"
+            },
+            {
+              "name": "large_explosion_particle",
+              "type": "particle"
+            },
+            {
+              "name": "explosion_sound_name",
+              "type": "string"
```
Comment (rom1504) [inline data/pc/1.20.3/protocol.json:2942]: this seems incorrect, see https://github.com/extremeheat/extracted_minecraft_data/blob/client1.20.3/client/net/minecraft/network/protocol/game/ClientboundExplodePacket.java#L121

File: data/pc/1.20.3/protocol.json [hunk 2]
```diff
@@ -0,0 +1,7309 @@
...
+                    "set_creative_slot": "packet_set_creative_slot",
+                    "update_jigsaw_block": "packet_update_jigsaw_block",
+                    "update_structure_block": "packet_update_structure_block",
+                    "update_sign": "packet_update_sign",
+                    "arm_animation": "packet_arm_animation",
+                    "spectate": "packet_spectate",
+                    "block_place": "packet_block_place",
+                    "use_item": "packet_use_item",
+                    "chunk_batch_received": "packet_chunk_batch_received",
+                    "configuation_acknowledged": "packet_configuation_acknowledged",
```
Comment (rom1504) [inline data/pc/1.20.3/protocol.json:7297]: typo

File: data/pc/1.20.3/protocol.json (hunk 1 above)
Comment (wgaylord) [reply data/pc/1.20.3/protocol.json:2942]: https://github.com/extremeheat/extracted_minecraft_data/blob/client1.20.3/client/net/minecraft/sounds/SoundEvent.java#L61-L64 

We don't have a soundEvent type so I just wrote what it has out (much like many other types for example the ClientOptions) and its also not used anywhere else that I can see,

File: data/pc/1.20.3/protocol.json (hunk 2 above)
Comment (wgaylord) [reply data/pc/1.20.3/protocol.json]: Fixed.

File: data/pc/1.20.3/protocol.json (hunk 1 above)
Comment (rom1504) [reply data/pc/1.20.3/protocol.json:2942]: yeah I see, it is indeed a string, they just parse it further but we don't need to

**Comment by rom1504** [comment]
ok lgtm thanks

**Pull request merged**